common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -531,6 +541,65 @@ protected SecretKey loadSecretKey(@NonNull final KeyType keyType) throws IOExcep
         throw new GeneralSecurityException(ErrorStrings.UNKNOWN_ERROR);
     }
 
+    /**
+     * If needed, get the key from inactive broker (if there is one).
+     * If it fails to get a new key. It will create a new one.
+     *
+     * This function can only be invoked by Broker apps.
+     */
+    public void migrateEncryptionKeyIfNeeded() throws GeneralSecurityException, IOException {
+        final String methodName = ":migrateEncryptionKeyIfNeeded";
+
+        final String activeBrokerPackageName = getPackageName();
+
+        if (!(COMPANY_PORTAL_APP_PACKAGE_NAME.equalsIgnoreCase(activeBrokerPackageName) ||
+                AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(activeBrokerPackageName))) {
+            final String errorMessage = "Caller is not a broker. Migration is not needed";
+            Logger.error(TAG + methodName, errorMessage, null);
+            throw new IllegalStateException(errorMessage);
+        }
+
+        mEncryptionKey = getKey();
+        if (mEncryptionKey != null){
+            Logger.verbose(TAG + methodName, "Key already exists.");
+            return;
+        }
+
+        mEncryptionKey = getKeyFromInactiveBroker(activeBrokerPackageName);
+        if (mEncryptionKey != null){
+            Logger.verbose(TAG + methodName, "Key is successfully retrieved from inactive broker. Saving key...");
+            saveSecretKey(mEncryptionKey);
+            return;
+        }
+
+        Logger.verbose(TAG + methodName, "Key migration failed. Create a new key.");
+        mEncryptionKey = createKey();
+        return;
+    }
+
+    @Nullable
+    protected SecretKey getKeyFromInactiveBroker(@NonNull final String activeBrokerPackageName) {
+        final String methodName = ":getKeyFromInactiveBroker";
+        final String serializedKey = InactiveBrokerClient.getSerializedSymmetricKeyFromInactiveBroker(mContext, activeBrokerPackageName);
+
+        if (serializedKey == null || serializedKey.length() == 0) {
+            Logger.verbose(TAG + methodName, "The returned bundle doesn't contain any key.");
+            return null;
+        }
+
+        return deserializeSecretKey(serializedKey);
+    }
+
+    private void saveSecretKey(@NonNull SecretKey key) throws GeneralSecurityException, IOException {
+        final String methodName = ":saveSecretKey";
+
+        Logger.verbose(TAG + methodName, "Saving secret key to storage.");
+
+        mKeyPair = generateKeyPairFromAndroidKeyStore();
```

Comment:
>mKeyPair [](start = 8, length = 8)

Shouldn't we check if mKeyPair is not null before generating a new keypair in Android key store?

Reply:
Technically it's not needed (given that this function will be called only when there is no existing secretKey (after it gets a key from IPC, or created a new one).

I'll add comments to clarify this.

Reply:
I want to keep the code that initializes member variables together. Otherwise, it is too hard to read.

---
In reply to: [301375521](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/547#discussion_r301375521) [](ancestors = 301375521)

Reply:
I decided to put the check as originally suggested to minimize the change - as we want to push this out asap.

It's hard to group initializer together because they are being generated in different places. In this case, the secretKey is obtained **from the other broker**, while this keypair is generated locally.

I was (in the process of) refactoring this layer to make it easier to read (i.e. extracting KeyStore and SecretKey wrapper with internal cache, meaning all this layer would have to do is to call wrapper.get())

However, that would require more change, and therefore might not be worth doing now.